internal/controller/driver_controller.go

```diff
@@ -18,45 +18,401 @@ package controller
+
+	// Extract the driver sort name and driver type
+	matches := nameRegExp.FindStringSubmatch(r.driver.Name)
+	if len(matches) != 3 {
+		return fmt.Errorf("invalid driver name")
+	}
+	r.driverName = matches[1]
+	r.driverType = strings.ToLower(matches[2])
+
+	// Load the current desired state in the form of a ceph csi driver resource
+	if err := r.Get(r.ctx, client.ObjectKeyFromObject(&r.driver), &r.driver); err != nil {
+		r.log.Error(err, "Unable to load driver.csi.ceph.io object", "name", client.ObjectKeyFromObject(&r.driver))
+		return client.IgnoreNotFound(err)
+	}
+
+	// Creating a copy of the driver spec, making sure any local changes will not effect the object residing
+	// in the client's cache
+	r.driver.Spec = *r.driver.Spec.DeepCopy()
+	mergeDriverSpecs(&r.driver.Spec, opConfig.Spec.DriverSpecDefaults)
+	mergeDriverSpecs(&r.driver.Spec, &driverDefaults)
+
+	r.images = maps.Clone(imageDefaults)
+	if r.driver.Spec.ImageSet != nil {
+		imageSetCM := corev1.ConfigMap{}
+		imageSetCM.Name = r.driver.Spec.ImageSet.Name
+		imageSetCM.Namespace = r.driver.Namespace
+		if err := r.Get(r.ctx, client.ObjectKeyFromObject(&imageSetCM), &imageSetCM); err != nil {
+			r.log.Error(err, "Unable to load driver specified image set config map", "name", client.ObjectKeyFromObject(&imageSetCM))
+			return err
+		}
+
+		maps.Copy(r.images, imageSetCM.Data)
+	}
+
+	return nil
+}
+
+func (r *driverReconcile) reconcileK8sCsiDriver() error {
+	existingCsiDriver := &storagev1.CSIDriver{}
+	existingCsiDriver.Name = r.driver.Name
+
+	log := r.log.WithValues("driverName", existingCsiDriver.Name)
+	log.Info("Reconciling CSI Driver resource")
+
+	if err := r.Get(r.ctx, client.ObjectKeyFromObject(existingCsiDriver), existingCsiDriver); client.IgnoreNotFound(err) != nil {
+		log.Error(err, "Failed to load CSI Driver resource")
+		return err
+	}
+
+	desiredCsiDriver := existingCsiDriver.DeepCopy()
+	desiredCsiDriver.Spec = storagev1.CSIDriverSpec{
+		AttachRequired: r.driver.Spec.AttachRequired,
+		PodInfoOnMount: ptr.To(false),
+		FSGroupPolicy:  &r.driver.Spec.FsGroupPolicy,
+	}
+
+	ownerObjKey := client.ObjectKeyFromObject(&r.driver)
+	if bytes, err := json.Marshal(ownerObjKey); err != nil {
+		log.Error(
+			err,
+			"Failed to JSON marshal owner obj key for CSI driver resource",
+			"ownerObjKey",
+			ownerObjKey,
+		)
+		return err
+	} else {
+		utils.AddAnnotation(desiredCsiDriver, ownerRefAnnotationKey, string(bytes))
+	}
+
+	if existingCsiDriver.UID == "" || !reflect.DeepEqual(desiredCsiDriver, existingCsiDriver) {
+		if existingCsiDriver.UID != "" {
+			log.Info("CSI Driver resource exist but does not meet desired state")
+			if err := r.Delete(r.ctx, existingCsiDriver); err != nil {
+				log.Error(err, "Failed to delete existing CSI Driver resource")
+				return err
+			}
+			log.Info("CSI Driver resource deleted successfully")
+		} else {
+			log.Info("CSI Driver resource does not exist")
+		}
+
+		if err := r.Create(r.ctx, desiredCsiDriver); err != nil {
```

Comment:
I assume this means it'll be essentially recreating the driver for all cases and that the nested conditions above were purposely written to allow it. Why not use separate `Update` and `Create` in this context? Is it the same argument you've made in my PR?

Reply:
Yes, the CSI driver Spec is mostly immutable and cannot be updated. The only thing that can be updated is the metadata (labels and annotations). 
There is no reason to make this procedure much more complex (to read and maintain) just to support an update op in case a label or annotation was modified and we need to revert it.

Reply:
please check on this one, not all fields are immutable in the csiDriver object. we don't need to create if there is a mutable change

Reply:
I will double-check but, when considering optimizing for a better fitting code vs optimizing for simplicity, where the update rate is very very very low and the side effects are almost non-existent, I will always choose to optimize for simplicity

Reply:
opening tracker to address this enhancement.